hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/transfer/AssociateTokenRecipientsStep.java

```diff
@@ -54,57 +63,76 @@ public void doIn(@NonNull final TransferContext transferContext) {
                 final var receiverId = aa.receiverAccountID();
                 final var senderId = aa.senderAccountID();
                 // sender should be associated already. If not throw exception
                 validateTrue(tokenRelStore.get(senderId, tokenId) != null, TOKEN_NOT_ASSOCIATED_TO_ACCOUNT);
-                validateAndAutoAssociate(
-                        receiverId, tokenId, token, accountStore, tokenRelStore, handleContext, recordBuilder);
+
+                final var nft = nftStore.get(tokenId, aa.serialNumber());
+                validateTrue(nft != null, INVALID_NFT_ID);
+
+                if (nft.hasOwnerId()) {
+                    validateTrue(nft.ownerId().equals(senderId), SENDER_DOES_NOT_OWN_NFT_SERIAL_NO);
+                } else {
+                    validateTrue(token.treasuryAccountId().equals(senderId), SENDER_DOES_NOT_OWN_NFT_SERIAL_NO);
+                }
+
+                validateAndBuildAutoAssociation(
+                        receiverId, tokenId, token, accountStore, tokenRelStore, handleContext, newAssociations);
             }
         }
+
+        for (TokenAssociation newAssociation : newAssociations) {
+            recordBuilder.addAutomaticTokenAssociation(newAssociation);
+        }
     }
 
     /**
      * Associates the token with the account if it is not already associated. It is auto-associated only if there are
      * open auto-associations available on the account.
      *
-     * @param accountId     The account to associate the token with
-     * @param tokenId       The tokenID of the token to associate with the account
-     * @param token         The token to associate with the account
-     * @param accountStore  The account store
-     * @param tokenRelStore The token relation store
-     * @param handleContext The context
-     * @param recordBuilder The record builder
+     * @param accountId         The account to associate the token with
+     * @param tokenId           The tokenID of the token to associate with the account
+     * @param token             The token to associate with the account
+     * @param accountStore      The account store
+     * @param tokenRelStore     The token relation store
+     * @param handleContext     The context
+     * @param newAssociations   The list of new associations to be created after all validations
      */
-    private void validateAndAutoAssociate(
+    private void validateAndBuildAutoAssociation(
             @NonNull final AccountID accountId,
             @NonNull final TokenID tokenId,
             @NonNull final Token token,
             @NonNull final WritableAccountStore accountStore,
             @NonNull final WritableTokenRelationStore tokenRelStore,
             @NonNull final HandleContext handleContext,
-            @NonNull final CryptoTransferRecordBuilder recordBuilder) {
+            @NonNull final List<TokenAssociation> newAssociations) {
```

Comment:
Nit: it would be nice to return a `TokenRelation` from this method and add the new tokenRel to `newAssociations` where this method is called, so that we don't have to pass in a mutable data structure here